### PR TITLE
Weather information and visitor counter

### PR DESCRIPTION
**Pull request by sherrero96**
I have added to the main page a script to show the zaragoza weather in the next days, with the help of:
[weatherwidget.io](https://weatherwidget.io/).
In addition I have added a counter of visits that makes a user to the web and a mini test to check that the number of visits is not negative.
I finally added some documentation. I explained in a simple way Spring Boot, how to create a project (already explained in the presentation of the course) and some java annotations.

### CODE REVIEW

File: description.md [hunk 1]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
 Refer to [Apache Tomcat documentation](https://tomcat.apache.org/tomcat-8.0-doc/deployer-howto.html) about how to deploy a WAR file, once [deliverables have been built](#how-to-build-the-code).
 
 ## Which are the technologies used in the code
```
Comment (fjlopez) [inline description.md]: Potential class with #27. I will warn you when #27 is merged.

File: description.md [hunk 2]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+### Spring
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 3]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+Spring Boot makes it easy to create stand-alone, production-grade Spring based Applications that you can "just run".
+
+We take an opinionated view of the Spring platform and third-party libraries so you can get started with minimum fuss. Most Spring Boot applications need very little Spring configuration.
+
+In fact, a professor of Web Engineering demonstrated how to mount URL Shortener in less than 5 minutes with Spring Boot.
+
+#### Some advantages over using spring
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 4]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+* It provides Embedded HTTP servers like Tomcat, Jetty etc. to develop and test our web applications very easily.
+* It provides CLI (Command Line Interface) tool to develop and test Spring Boot(Java or Groovy) Applications from command prompt very easily and quickly.
+* And much more.
+
+The sources used and where much more information can be found:
+* [spring.io](https://spring.io/projects/spring-boot)
+* [hwww.journaldev.com/7969/spring-boot-tutorial](https://www.journaldev.com/7969/spring-boot-tutorial)
 
 ## How these technologies work
+### To start a project with Spring boot
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 5]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+#### Prerequisites
+1. We are going to use the gradle tool, so it must be installed.
+
+#### Create Spring Boot proyect
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 6]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+implementation 'org.springframework.boot:spring-boot-starter-web'
+```
+4. This is the initial application of Spring Boot.
+
+5. Now you can add the actions that are triggered when making requests to endpoints.
 
 ## What means each a specific piece or code
 
 ## Which is the purpose of a specific Java annotation
+### @Controller
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 7]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+As spring boot it follows the architecture Model-View-Controller, @Controller annotation indicates that the annotated class is a controller. It is a specialization of @Component and is autodetected through classpath scanning. It is typically used in combination with annotated handler methods based on the @RequestMapping annotation.
+In short, it serves to indicate that it is the controller class of the architecture.
+
+### @Value
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 8]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+
+The @Value annotation is used to read the value of some of the properties stored in the properties file.
+In this file, you can see data about the environment, about values or variables or even properties of the application as the listening port.
+The syntax of the @Value annotation is
+```java
+@Value("${property_key_name}")
+```
+
+
+### @GetMapping("/")
```
Comment (fjlopez) [inline description.md]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md022)

File: description.md [hunk 9]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+* [spring.io](https://spring.io/projects/spring-boot)
+* [hwww.journaldev.com/7969/spring-boot-tutorial](https://www.journaldev.com/7969/spring-boot-tutorial)
 
 ## How these technologies work
+### To start a project with Spring boot
+#### Prerequisites
+1. We are going to use the gradle tool, so it must be installed.
+
+#### Create Spring Boot proyect
+1. Create gradle proyect
```
Comment (fjlopez) [inline description.md]: `project`

File: description.md [hunk 10]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+```bash
```
Comment (fjlopez) [inline description.md]: [Fenced code blocks should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md031)

File: description.md [hunk 11]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+gradle init
+```
+Let's select the desired options, using java as language.
+2. Transform into a Spring Boot Web application
+Edit the class App at src/... and adds the following imports:
+```java
```
Comment (fjlopez) [inline description.md]: [Fenced code blocks should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md031)

File: description.md [hunk 12]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+```
+Above the definition of the App function adds : @SpringBootApplication
+And finally, within the main function write SpringApplication.run(App.class, args);
+3. Editing the gradle builder
+Add the following plugins and dependencies:
+```java
```
Comment (fjlopez) [inline description.md]: [Fenced code blocks should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md031)

File: description.md [hunk 13]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+#### Prerequisites
+1. We are going to use the gradle tool, so it must be installed.
+
+#### Create Spring Boot proyect
+1. Create gradle proyect
+```bash
+gradle init
+```
+Let's select the desired options, using java as language.
+2. Transform into a Spring Boot Web application
```
Comment (fjlopez) [inline description.md]: Add a blank line before it.

File: description.md [hunk 14]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+Edit the class App at src/... and adds the following imports:
+```java
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+```
+Above the definition of the App function adds : @SpringBootApplication
+And finally, within the main function write SpringApplication.run(App.class, args);
+3. Editing the gradle builder
```
Comment (fjlopez) [inline description.md]: Add a blank line before it.

File: description.md (hunk 5 above)
Comment (fjlopez) [inline description.md]: Lists in markdown are a bit tricky if they are multilines. 

#### This is your markdown
1. Create gradle proyect
```bash
gradle init
```
Let's select the desired options, using java as language.
2. Transform into a Spring Boot Web application
#### This is your markdown after the fixes
1. Create gradle proyect

    ```bash
    gradle init
    ```

    Let's select the desired options, using java as language.

1. Transform into a Spring Boot Web application

This is the code
```markdown
1. Create gradle proyect

    ```bash
    gradle init
    ```

    Let's select the desired options, using java as language.

1. Transform into a Spring Boot Web application
```
Two main advantages:
- automatic numbering (this is optional, you can keep numbering)
- legibility (easiness to identify blocks without rendering the markdown)

File: description.md [hunk 15]
```diff
@@ -48,11 +48,82 @@ gradle bootRun
...
+```
+Let's select the desired options, using java as language.
+2. Transform into a Spring Boot Web application
+Edit the class App at src/... and adds the following imports:
+```java
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+```
+Above the definition of the App function adds : @SpringBootApplication
+And finally, within the main function write SpringApplication.run(App.class, args);
```
Comment (fjlopez) [inline description.md]: Use backticks for fencing inlined code, that is:

```markdown
And finally, within the main function write `SpringApplication.run(App.class, args);`
```

This is the result:

----

And finally, within the main function write `SpringApplication.run(App.class, args);`

File: src/main/java/es/unizar/webeng/hello/HelloController.java [hunk 16]
```diff
@@ -82,6 +85,11 @@ public String welcome(Map<String, Object> model) {
             logger.debug("Request object is not valid");
         }
         model.put("message", message);
+
+        /* It adds one to the visits of the web */
+        visitorCount++;
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/HelloController.java]: Web applications are asynchronous per nature. The use of `int` as a shared objet will cause race conditions if its value is modified as is. Code must be thread-safe. There are several ways of rewrite this code and fix this issue. Implement one of them.

File: src/test/java/es/unizar/webeng/hello/HelloControllerUnitTest.java [hunk 17]
```diff
@@ -93,4 +94,16 @@ public void testClient() throws Exception {
...
+    @Test
+    public void testCount() throws Exception{
+        HashMap<String, Object> map = new HashMap<>();
+        String view = controller.welcome(map);
+        /** Checks if controller uses "wellcome" view */
+        assertThat(view, is("wellcome"));
+        /**Checks if data exist and is valid */
+        assertThat(map.containsKey("visitorCount"), is(true));
+        /** Visits can't be negative */
+        assertTrue("Visitor count should be greater or equal to 0", (int)map.get("visitorCount") >= 0);
```
Comment (fjlopez) [inline src/test/java/es/unizar/webeng/hello/HelloControllerUnitTest.java]: Must be greater than 0 as you have already visited the controller.

**Comment by sherrero96** [comment]
I think I've fixed all the requested changes

File: description.md [hunk 18]
```diff
@@ -49,10 +49,101 @@ Refer to [Apache Tomcat documentation](https://tomcat.apache.org/tomcat-8.0-doc/
...
+1. Transform into a Spring Boot Web application
+    Edit the class App at src/... and adds the following imports:
+
+    ```java
+    import org.springframework.boot.SpringApplication;
+    import org.springframework.boot.autoconfigure.SpringBootApplication;
+    ```
+
+    Above the definition of the App function adds : @SpringBootApplication
+    And finally, within the main function write `SpringApplication.run(App.class, args);
```
Comment (fjlopez) [inline description.md]: A closing backtick is missing here.

File: src/test/java/es/unizar/webeng/hello/HelloControllerUnitTest.java [hunk 19]
```diff
@@ -93,4 +94,15 @@ public void testClient() throws Exception {
...
+
+    @Test
+    public void testCount() throws Exception{
+        HashMap<String, Object> map = new HashMap<>();
+        String view = controller.welcome(map);
+        /** Checks if controller uses "wellcome" view */
+        assertThat(view, is("wellcome"));
+        /**Checks if data exist and is valid */
+        assertThat(map.containsKey("visitorCount"), is(true));
+        assertNotNull(map.get("visitorCount"));
```
Comment (fjlopez) [inline src/test/java/es/unizar/webeng/hello/HelloControllerUnitTest.java]: Test with some value. If your code works, the value of visitor count must be greater than 0.

**Comment by sherrero96** [comment]
I have fixed the missing backtick and added a value in the class test as you have requested.

**Comment by fjlopez** [comment]
Waiting to #27. Be ready to potential conflicting changes but in any scenario you have won a :gift:.

**Comment by fjlopez** [comment]
#27 has added his code and it has generated conflicting files. You should resolve the conflicts locally and push the resolution to your repo.

**Comment by sherrero96** [comment]
I think I've already fixed the conflicts.

**Comment by fjlopez** [comment]
Con :gift:

**Pull request merged**